keps/sig-api-machinery/3693-admission-webhook-exclusion-rules/README.md

```diff
@@ -0,0 +1,944 @@
+-->
+
+## Design Details
+
+Sample ValidatingWebhookConfiguration
+```
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: "test"
+webhooks:
+  - name: "pod-policy.example.com"
+    rules:
+      - apiGroups:   [""]
+        apiVersions: ["v1"]
+        operations:  ["CREATE"]
+        resources:   ["pods"]
+        scope: "Namespaced"
+    clientConfig: # omitted
+    admissionReviewVersions: ["v1"]
+    sideEffects: None
+    timeoutSeconds: 5
+    exclusionRules: # here is the new field
+      - apiGroups:   [""] #should support wildcard just like regular rules
+        apiVersions: ["v1"]
+        operations:  ["CREATE"]
+        resources:   ["pods"]
+        objectNames: ["test"]
+        scope: "Namespaced"
+        Namespaces: ["test"]
+# name and namespace because user more likely to want more specificity with exclusion rules.
+```
+We have a set of exclusion rules per object, if any of the rules match the request, the webhook will not intercept the request. 
+
+**API Sketch**
+```
+type ValidatingWebhook struct {
+  ...
+	// ExcludeRules describes what operations on what resources/subresources the ValidatingWebhook should ignore.
+	// The exclude rules take precedence over include rules (if a resource matches both, it is excluded)
+	// +optional
+	ExcludeRules []NamedRuleWithOperations
+}
+
+// NamedRuleWithOperations is a tuple of Operations and Resources with ResourceNames.
+// +structType=atomic
+type NamedRuleWithOperations struct {
+	// ResourceNames is an optional list of object names that this rule applies to.  An empty list means that all names are matched.
+	// +optional
+	ObjectNames []string
+	// Namespaces is an optional list of namespaces that this rule applies to. An empty list means that all namespaces (and cluster scoped resources) are matched.
+	// +optional
+	Namespaces []string  # Namespace selector or specific list of namespace names?
+	// RuleWithOperations is a tuple of Operations and Resources.
+	RuleWithOperations
+}
+
+```
+
+```
+<<[UNRESOLVED @tallclair ]>>
+namespaces list or namespace selector?
+<<[/UNRESOLVED]>>
+```
+
+```
+<<[UNRESOLVED @tallclair ]>>
+@tallclair: I wonder if we want to streamline this API a bit more? I think it will be common to define the matching rules like:
+
+rules:
+- apiGroups: [""] // core
+apiVersions: ["*"]
+operations: ["create", "update"]
+resources: ["deployments"]
+
+And then want to exclude a few resource explicitly:
+
+exclusionRules:
+- objectNames: ["monitoring-server", "kube-proxy"]
+namespaces: ["kube-system"]
+
+In this case, I shouldn't need to repeat the apiGroups, apiVersions, operations & resources from the match rules. Leaving them blank should implicitly match everything.
+
+@ivelichkovich: So how does that work if the user lists multiple rules though? how we would we know which set of the apiGroups/apiVersions/operations from which rule we want to inherit values from?
```

Comment:
It's not really inheritance. I'm just suggesting treating empty value the same as `"*"`. We wanted to make `*` an explicit opt-in for the match rules, but I'm not sure the same makes sense for exclusion rules.

Reply:
Oh I thought you meant explicitly matching the previous rule, that makes more sense.